File: articles/machine-learning-life-sciences/julia/ml_life_sciences_kernel.jl
# Numerical machine-learning kernel for life-science classification in Julia.
# This dependency-free example demonstrates logistic scoring.

sigmoid(z) = 1.0 / (1.0 + exp(-z))

function predict_probability(immune, metabolic, morphology, stress)
    linear_score = 3.0 * immune - 2.0 * metabolic + 2.2 * morphology + 1.2 * stress - 2.0
    return sigmoid(linear_score)
end

samples = [
    ("EXT001", 0.82, 0.22, 0.76, 0.71, 1),
    ("EXT002", 0.74, 0.29, 0.70, 0.67, 1),
    ("EXT003", 0.38, 0.64, 0.39, 0.41, 0),
    ("EXT004", 0.27, 0.71, 0.33, 0.36, 0),
    ("EXT005", 0.63, 0.36, 0.60, 0.59, 1),
    ("EXT006", 0.43, 0.57, 0.45, 0.43, 0),
]

correct = 0

for sample in samples
    id, immune, metabolic, morphology, stress, observed = sample
    p = predict_probability(immune, metabolic, morphology, stress)
    predicted = p >= 0.5 ? 1 : 0
    global correct += predicted == observed ? 1 : 0
    println(id, " probability=", round(p, digits=5), " predicted=", predicted, " observed=", observed)
end

println("accuracy=", round(correct / length(samples), digits=5))

File: articles/machine-learning-life-sciences/fortran/ml_life_sciences_kernel.f90
program ml_life_sciences_kernel
  implicit none

  integer, parameter :: n = 6
  integer :: i, observed(n), predicted, correct
  real :: immune(n), metabolic(n), morphology(n), stress(n), probability

  immune = (/0.82, 0.74, 0.38, 0.27, 0.63, 0.43/)
  metabolic = (/0.22, 0.29, 0.64, 0.71, 0.36, 0.57/)
  morphology = (/0.76, 0.70, 0.39, 0.33, 0.60, 0.45/)
  stress = (/0.71, 0.67, 0.41, 0.36, 0.59, 0.43/)
  observed = (/1, 1, 0, 0, 1, 0/)

  correct = 0

  do i = 1, n
     probability = sigmoid(3.0 * immune(i) - 2.0 * metabolic(i) + 2.2 * morphology(i) + 1.2 * stress(i) - 2.0)

     if (probability >= 0.5) then
        predicted = 1
     else
        predicted = 0
     end if

     if (predicted == observed(i)) correct = correct + 1

     print *, "sample=", i, " probability=", probability, " predicted=", predicted, " observed=", observed(i)
  end do

  print *, "accuracy=", real(correct) / real(n)

contains

  real function sigmoid(z)
    real, intent(in) :: z
    sigmoid = 1.0 / (1.0 + exp(-z))
  end function sigmoid

end program ml_life_sciences_kernel
